### PR TITLE
Narrow on element access of known property

### LINKED ISSUE #10530: Discriminant property type guard not applied with bracket notation
**TypeScript Version:**  2.0.0

**Code**

``` ts
interface Square {
    kind: "square";
    size: number;
}

interface Rectangle {
    kind: "rectangle";
    width: number;
    height: number;
}

interface Circle {
    kind: "circle";
    radius: number;
}

type Shape = Square | Rectangle | Circle;

function area(s: Shape) {
    // In the following switch statement, the type of s is narrowed in each case clause
    // according to the value of the discriminant property, thus allowing the other properties
    // of that variant to be accessed without a type assertion.
    switch (s['kind']) {
        case "square": return s.size * s.size;
        case "rectangle": return s.width * s.height;
        case "circle": return Math.PI * s.radius * s.radius;
    }
}
```

**Expected behavior:**

The code compiles without errors.

**Actual behavior:** 

```
sample.ts(24,33): error TS2339: Property 'size' does not exist on type 'Square | Rectangle | Circle'.
sample.ts(24,42): error TS2339: Property 'size' does not exist on type 'Square | Rectangle | Circle'.
sample.ts(25,36): error TS2339: Property 'width' does not exist on type 'Square | Rectangle | Circle'.
sample.ts(25,46): error TS2339: Property 'height' does not exist on type 'Square | Rectangle | Circle'.
sample.ts(26,43): error TS2339: Property 'radius' does not exist on type 'Square | Rectangle | Circle'.
sample.ts(26,54): error TS2339: Property 'radius' does not exist on type 'Square | Rectangle | Circle'.
```

**Why this is bad:**

I am trying to work with Dropbox's new 2.0 SDK, which _heavily_ uses tagged union types (especially for API errors). The discriminant property is named `.tag`, so it can only be accessed via bracket notation. I generated TypeScript typings for their new JavaScript SDK, and discovered this bug the hard way. :(

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #10530 

Narrowing now happens for `x['knownProperty']` the same way that it does for `x.knownProperty`

### CODE REVIEW

**Comment by sandersn** [comment]
@RyanCavanaugh @ahejlsberg do you want to take a look?

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -608,9 +609,21 @@ namespace ts {
         }
 
         function isNarrowableReference(expr: Expression): boolean {
-            return expr.kind === SyntaxKind.Identifier ||
-                expr.kind === SyntaxKind.ThisKeyword ||
-                expr.kind === SyntaxKind.PropertyAccessExpression && isNarrowableReference((<PropertyAccessExpression>expr).expression);
+            if (expr.kind === SyntaxKind.ElementAccessExpression) {
+                const argument = (expr as ElementAccessExpression).argumentExpression;
+                return argument.kind === SyntaxKind.StringLiteral || argument.kind === SyntaxKind.NumericLiteral || argument.kind === SyntaxKind.EnumMember;
```
Comment (mhegazy) [inline src/compiler/binder.ts]: the expr.expression should be, an identifier, this, a PopertyAccessExpression or ElementAccessExpression , so we should recursively call the function on the expression.

Comment (mhegazy) [inline src/compiler/binder.ts]: we should also add tests to multiple levels of elementAccess. `a[0][1]`

Comment (sandersn) [inline src/compiler/binder.ts]: Actually, multiple levels of property access don't work either. So I think this function should actually be even simpler until we fix that bug.

```ts
interface Square {
  sub: { kind: "square; };
  size: number;
}
// etc ...
function area(s: Shape) {
    switch(s.sub.kind) {
        case "square": return s.size * s.size;
        case "rectangle": return s.width * s.height;
        case "circle": return Math.PI * s.radius * s.radius;
    }
}
```

**Comment by sandersn** [comment]
Nested element accesses are now supported.

**Comment by sandersn** [comment]
@mhegazy you started reviewing this change. Can you look at the new commits so it can go in?

**Comment by sandersn** [comment]
I ran a performance benchmark on the code compared to master. The standout difference was that Monaco (the online component that powers the playground) has a 6.85% slowdown in the checker, or 0.41 second. TFS has a small slowdown in the checker but it's not statistically significant. 

I'm trying to get github formatting to co-operate so I can show the actual table but haven't had any luck yet.

**Comment by sandersn** [comment]
As @mhegazy suggested, I tried the benchmark again, this time only narrowing string element accesses, not number accesses. The results are about the same.

**Comment by mhegazy** [comment]
given the performance penalty vs. utility of this change, we will not be able to take it.

**Comment by kachkaev** [comment]
Any chance this can be revised 1.5 years since the original PR? Would the performance impact be so critical now too?

**Comment by mhegazy** [comment]
i do not think the error you listed above has do with this PR. The issue here is that it is legal to index into any property using a string/numeric name. the result is an implicit `any`. This should be an error under `--noImplicitAny`.

**Comment by kachkaev** [comment]
I'm not sure that `--noImplicitAny` would work for me here, because I'll have to add type annotation when defining `config` and the code will become too verbose. `--noImplicitAny` also affects function argument definition, which I'd prefer to avoid.

**Comment by m1gu3l** [comment]
could this perhaps land behind a flag?

**Comment by sandersn** [comment]
I reconstituted the PR and found (1) a probably performance bug that it introduced in `isMatchingReference` and friends (2) no significant difference in performance after fixing the bug. I'll open a new PR with the details and some tests.

**Comment by BalaM314** [comment]
Can we have this behind a flag? I've had to add dozens of non-null assertions to fix this and would gladly take the 6.5% performance hit.

**Comment by thelinuxlich** [comment]
I agree, the current behavior is unintuitive and should be fixed.

**Comment by phaseOne** [comment]
For what it's worth, I've verified that `noImplicitAny`/`strict` addresses the case in https://github.com/microsoft/TypeScript/pull/10565#issuecomment-376530160.

For my use case, `strict` is fine to enable. The only reason I'm making this post is that it's unintuitive that `noImplicitAny` controls type checking for property accessors (bracket notation) in common cases like the aforementioned, so hopefully, this post enables the algorithm to help other users find the workaround and latest news.

It's purported that #10530 is tracking the issue and #48335 is tracking new documentation for the current behavior and workarounds.

**Comment by Zzzen** [comment]
Would it be possible for us to simply rebase this branch and rerun the benchmarks to get a clearer picture of the real-world performance?

**Comment by DamianGlowala** [comment]
@sandersn, what do you think about the above question? Could we give this another try, please?

**Comment by RyanCavanaugh** [comment]
One does not "simply" rebase a 7-year-old PR 😅

The example in the linked issue, and the example in this comment https://github.com/microsoft/TypeScript/pull/10565#issuecomment-397276283, both now work. Can you clarify what you're looking for?

**Comment by DamianGlowala** [comment]
Now as I am reading the purpose of this PR again, I certainly commented on the wrong one. I was trying to figure out why TS doesn't narrow down the type [in this specific scenario](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBCBGArAXDASgU1AJwCYB5psBLMAcwBoYjSyA+GAXhgG8BfAKA9ElgAdsIPhmxQAnkxgAKKQEomDKQHIA1hjFLZsuV2IAzaQKEjxMUnCTyWHGDB7QYAfUkJEAbSPDRYgLowA9P5OZhAwAK5guBh6pBi4MADuxFAAFjBKYCAAqmDAKVhquACSkRgAHnEAgsDAGBAQqFDYYRhKHGxAA0).

**Pull request closed without merging**